compile error fixes

cscipoll opened the pull request:
-IPieceState made into a script to work with ChessStates -made (byte)ChessPieceData.firstMove unchecked in ChessData.cs -changed ToStdring() to ToString() on DragAndDrop.cs

cscipoll committed:
compile error fixes
-IPieceState made into a script to work with ChessStates
-made (byte)ChessPieceData.firstMove unchecked in ChessData.cs
-changed ToStdring() to ToString() on DragAndDrop.cs

MichaelCipolla commented on Assets/IPieceState.cs:
```diff
@@ -0,0 +1,18 @@
+using System.Collections;
```
Ok, so if IPieceState was never defined we're going to need to define the interface correctly. Interfaces differ from classes in the following ways:

MichaelCipolla commented on Assets/IPieceState.cs:
```diff
@@ -0,0 +1,18 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class IPieceState : MonoBehaviour
```
Instead of class, let's change this to an interface with the interface keyword:
`public interface IPieceState`

MonoBehaviour is a specific Unity class that is used for elements that are accessible from Unity. Knowing when to extend from monobehaviour is an advanced kind of topic, but for the sake of making an interface, they typically don't extend from a parent class.

Here's an MSDN document on interfaces:
https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/types/interfaces

MichaelCipolla commented on Assets/IPieceState.cs:
```diff
@@ -0,0 +1,18 @@
+public class IPieceState : MonoBehaviour
+{
+    // Start is called before the first frame update
+    void Start()
```
void Start() and void Update() are methods of MonoBehavior. Since we aren't planning on extending from monobehavior we can remove them. What we do want to have here is a method header for our validatePiece() method that our piece states are extending. Go ahead and add that method header to this file.

MichaelCipolla commented on Assets/ChessStates.cs:
```diff
@@ -14,14 +14,18 @@ boolean validateMove(int startIndex, int endIndex)
 
         */
-        if(endIndex >= startIndex + 7 && endIndex <= startIndex + 9)
-        return false;
```
Ok so for the pawn movement, I think we have the logic a little messed up here. Wasn't intending for this section to have a bug but when considering the chess board, this is how the piece movement works:

We have an 8x8 board where each matrix cell has a number from 0 -> 63.

The startIndex is the index the piece begins on, and the endIndex is where the piece is placed.

We know that pawns can move diagonally 1 square in the left and right direction, as well as up 1 square in the forward direction.

If we were to imagine for instance that our pawn starts on square 3, his valid moves would be the square immediately on top of him, and the square to the right and left diagonally. Mathematically this works out to startIndex +7 (top left) startIndex + 8 (top) startIndex +9 (top right).

If our pawn's endIndex falls in that range, we want to return true, not false. For now, let's just validate the moves for whichever piece color starts on the bottom. We can work on distinguishing the piece colors later on.

MichaelCipolla reviewed (changes requested):
Good job fixing those errors. I left some comments on some changes I would like you to make before we merge this into main.

cscipoll committed:
compilation error edit v2
-Made IPieceState an interface and added the validatePiece() header
-fixed the true/false return values for pawn moves in ChessStates

MichaelCipolla reviewed (approved):
Alright. Looks good to merge.
